clients/venice-admin-tool/src/main/java/com/linkedin/venice/datarecovery/DataRecoveryModule.java

```diff
@@ -0,0 +1,102 @@
+package com.linkedin.venice.datarecovery;
+
+import java.io.Console;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * Module is the engine to run Tasks in data recovery.
+ */
+public class DataRecoveryModule {
+  private final Logger LOGGER = LogManager.getLogger(DataRecoveryModule.class);
+  private final static int DEFAULT_POOL_SIZE = 10;
+  private final static int DEFAULT_POOL_TIMEOUT_IN_SECONDS = 30;
+  private final int poolSize;
+  private final ExecutorService pool;
+  private List<DataRecoveryTask> tasks;
+
+  public DataRecoveryModule() {
+    this(DEFAULT_POOL_SIZE);
+  }
+
+  public DataRecoveryModule(int poolSize) {
+    this.poolSize = poolSize;
+    this.pool = Executors.newFixedThreadPool(this.poolSize);
+  }
+
+  public void perform(Set<String> storeNames, StoreRepushCommand.Params params) {
+    String pass = getUserCredentials();
+    if (pass == null) {
+      LOGGER.error("Cannot get password, exit");
+      return;
+    }
+    params.setPassword(pass);
+
+    tasks = buildTasks(storeNames, params);
+    List<CompletableFuture<Void>> taskFutures = tasks.stream()
+        .map(dataRecoveryTask -> CompletableFuture.runAsync(dataRecoveryTask, pool))
+        .collect(Collectors.toList());
+    taskFutures.stream().map(CompletableFuture::join).collect(Collectors.toList());
+    displayTaskResult();
+
+    shutdownAndAwaitTermination();
+  }
+
+  private void shutdownAndAwaitTermination() {
+    pool.shutdown();
+    try {
+      if (!pool.awaitTermination(DEFAULT_POOL_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS)) {
+        // Cancel currently executing tasks.
+        pool.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      // (Re-)Cancel if current thread also interrupted.
+      pool.shutdownNow();
+    }
+  }
+
+  public List<DataRecoveryTask> buildTasks(Set<String> storeNames, StoreRepushCommand.Params params) {
+    List<DataRecoveryTask> tasks = new ArrayList<>();
+    for (String name: storeNames) {
+      DataRecoveryTask.TaskParams taskParams = new DataRecoveryTask.TaskParams(name, params);
+      tasks.add(new DataRecoveryTask(taskParams));
+    }
+    return tasks;
+  }
+
+  public String getUserCredentials() {
+    Console console = System.console();
+    if (console == null) {
+      LOGGER.warn("System.console is null");
+      return null;
+    }
+    // Read password into character array.
+    char[] passwordVip = console.readPassword("Enter Credentials: ");
```

Comment:
Do we need this password+vip for every task execution or it will only be used once?
If it is required for each execution, this strategy might not be good enough since `vip` will change periodically if each task will take a long time since the executor thread pool is limited.

Reply:
In this implementation, it asks for the credential once, and then the input credentials among all tasks triggered later without the need to input it for every store. You are right on the point that for this version of implementation, vip can change periodically and there is a risk of having an expired vip, there is away to fully overcome this issue by leveraging the Azkaban headless challenge authentication. If the data recovery idempotent is ready, it would alleviate this issue a bit by rerunning the same command.